### PR TITLE
apply patterns like glob.glob

### PR DESCRIPTION
**Pull request by cscetbon**
Fixing issue https://github.com/dask/s3fs/issues/125

### CODE REVIEW

File: s3fs/core.py [hunk 1]
```diff
@@ -645,10 +645,10 @@ def glob(self, path):
         # must be a / between the bucket name and the *.
         ind = path[:path.index('*')].rindex('/')
         root = path[:ind + 1]
-        allfiles = self.walk(root)
+        allfiles = self.walk(root, directories=True)
 
         # Translate the glob pattern into regex (similar to `fnmatch`).
-        regex_text = '^'
+        regex_text = ''
```
[the remark this answers is not in the document]
Comment (cscetbon) [reply s3fs/core.py]: @martindurant You don't need this. `match` matches at the beginning of the string passed.

File: s3fs/core.py [hunk 2]
```diff
@@ -657,9 +657,11 @@ def glob(self, path):
             else:
                 regex_text += re.escape(c)
         regex_text += '$'
+
         regex_pattern = re.compile(regex_text)
 
-        out = [f for f in allfiles if regex_pattern.match(f)]
+        out = set([f for f in allfiles for m in [regex_pattern.match(f)] if m])
```
[the remark this answers is not in the document]
Comment (cscetbon) [reply s3fs/core.py]: @martindurant, the extra loop is no longer necessary. I had previous code that was using it but that I did not commit.
Okay I'm gonna change it a bit to keep the order

File: s3fs/core.py [hunk 3]
```diff
@@ -594,11 +594,11 @@ def _walk(self, path, refresh=False):
             if f['StorageClass'] == 'DIRECTORY':
                 filenames.extend(self._walk(f['Key'], refresh))
         return [f for f in filenames if f['StorageClass'] not in
-                ['BUCKET', 'DIRECTORY']]
+                (['BUCKET'] + ['DIRECTORY'] if not directories else [])]
```
[the remark this answers is not in the document]
Comment (cscetbon) [reply s3fs/core.py]: The current code returns `['BUCKET', 'DIRECTORY']` or `['BUCKET']` if `directories` is set. Isn't it what we want ?

File: s3fs/core.py [hunk 4]
```diff
@@ -657,12 +657,11 @@ def glob(self, path):
...
         regex_text += '$'
+
         regex_pattern = re.compile(regex_text)
 
-        out = [f for f in allfiles if regex_pattern.match(f)]
-        if not out:
-            out = self.ls(path0)
-        return out
+        seen = set()
+        return [f for f in allfiles if regex_pattern.match(f) and not (f in seen or seen.add(f))]
```
Comment (martindurant) [inline s3fs/core.py]: This is inventive

**Comment by martindurant** [comment]
Thank you, looks good.

**Comment by cscetbon** [comment]
One more fix https://github.com/dask/s3fs/pull/129

**Pull request merged**